totp/totp.go

```diff
@@ -74,13 +71,55 @@ type ValidateOpts struct {
 	Algorithm otp.Algorithm
 }
 
+func (opts *ValidateOpts) defaultOpts() {
+	if opts.Skew == 0 {
+		opts.Skew = 30
+	}
+	if opts.Digits == 0 {
+		opts.Digits = otp.DigitsSix
+	}
+	if opts.Period == 0 {
+		opts.Period = 30
+	}
+}
+
+type ValidateOpt func(opt *ValidateOpts)
+
+func WithPeriod(period uint) ValidateOpt {
+	return func(opt *ValidateOpts) {
+		opt.Period = period
+	}
+}
+func WithSkew(skew uint) ValidateOpt {
+	return func(opt *ValidateOpts) {
+		opt.Skew = skew
+	}
+}
+
+func WithDigits(digits otp.Digits) ValidateOpt {
+	return func(opt *ValidateOpts) {
+		opt.Digits = digits
+	}
+}
+
+func WithAlgorithm(algo otp.Algorithm) ValidateOpt {
+	return func(opt *ValidateOpts) {
+		opt.Algorithm = algo
+	}
+}
+
 // GenerateCodeCustom takes a timepoint and produces a passcode using a
 // secret and the provided opts. (Under the hood, this is making an adapted
 // call to hotp.GenerateCodeCustom)
-func GenerateCodeCustom(secret string, t time.Time, opts ValidateOpts) (passcode string, err error) {
-	if opts.Period == 0 {
-		opts.Period = 30
+func GenerateCodeCustom(secret string, t time.Time, validateOpts ...ValidateOpt) (passcode string, err error) {
```

Comment:
this breaks semver compat right?

can we introduce new functions so v1 remains API compataible?

Reply:
that's right, it breaks semver. We can introduce new funcs and add **deprecated** tags to the current ones, if possible.

Reply:
SGTM

Reply:
what's the next move? should I make any changes?

Reply:
I think the path forward is to make a new function, GenerateCodeCustomOpts, and label GenerateCodeCustom as depreciated, See GenerateCodeCustomOpts.

I think Generate, and ValidateCustom needs the same treatment.

I don't want to break ABI compatibility for this kind of change - its a good thing to do, but on the aesthetics side, so I think new funcs that don't break the API are the way to go.

Reply:
I'll work on it...